stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -0,0 +1,1375 @@
+                     parent_block_id TEXT NOT NULL,
+                     affirmation_weight INTEGER NOT NULL,
+                     -- this field is the total number of *tenures* in the chain history (including this tenure)
+                     tenure_height INTEGER NOT NULL,
+                     -- this field is true if this is the first block of a new tenure
+                     tenure_changed INTEGER NOT NULL,
+                     -- this field tracks the total tx fees so far in this tenure. it is a text-serialized u128
+                     tenure_tx_fees TEXT NOT NULL,
+              PRIMARY KEY(consensus_hash,block_hash)
+          );
+    "#.into(),
+        format!(
+            r#"ALTER TABLE payments
+               ADD COLUMN schedule_type TEXT NOT NULL DEFAULT "{}";
+            "#,
+            HeaderTypeNames::Epoch2.get_name_str()),
+        r#"
+        UPDATE db_config SET version = "4";
+        "#.into(),
+    ];
+}
+
+pub struct SetupBlockResult<'a, 'b> {
+    pub clarity_tx: ClarityTx<'a, 'b>,
+    pub tx_receipts: Vec<StacksTransactionReceipt>,
+    pub matured_miner_rewards_opt:
+        Option<(MinerReward, Vec<MinerReward>, MinerReward, MinerRewardInfo)>,
+    pub evaluated_epoch: StacksEpochId,
+    pub applied_epoch_transition: bool,
+    pub burn_stack_stx_ops: Vec<StackStxOp>,
+    pub burn_transfer_stx_ops: Vec<TransferStxOp>,
+    pub auto_unlock_events: Vec<StacksTransactionEvent>,
+    pub burn_delegate_stx_ops: Vec<DelegateStxOp>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct NakamotoBlockHeader {
+    pub version: u8,
+    /// The total number of StacksBlock and NakamotoBlocks preceding
+    /// this block in this block's history.
+    pub chain_length: u64,
+    /// Total amount of BTC spent producing the sortition that
+    /// selected this block's miner.
+    pub burn_spent: u64,
+    /// The block hash of the immediate parent of this block.
+    pub parent: BlockHeaderHash,
+    /// The bitcoin block whose data has been handled most recently by
+    /// the Stacks chain as of this block.
+    pub burn_view: BurnchainHeaderHash,
+    /// The root of a SHA512/256 merkle tree over all this block's
+    /// contained transactions
+    pub tx_merkle_root: Sha512Trunc256Sum,
+    /// The MARF trie root hash after this block has been processed
+    pub state_index_root: TrieHash,
+    /// Recoverable ECDSA signature from the tenure's miner.
+    pub signature: MessageSignature,
+}
+
+#[derive(Debug, Clone)]
+pub struct NakamotoBlock {
+    pub header: NakamotoBlockHeader,
+    pub txs: Vec<StacksTransaction>,
+}
+
+pub struct NakamotoChainState;
+
+impl FromRow<NakamotoBlockHeader> for NakamotoBlockHeader {
+    fn from_row(row: &rusqlite::Row) -> Result<NakamotoBlockHeader, DBError> {
+        let version = row.get("version")?;
+        let chain_length_i64: i64 = row.get("chain_length")?;
+        let chain_length = chain_length_i64.try_into().map_err(|_| DBError::Overflow)?;
```

Comment:
It's cleaner to use shadowing than create intermediate variables:

```suggestion
        let chain_length: i64 = row.get("chain_length")?;
        let chain_length = chain_length.try_into().map_err(|_| DBError::Overflow)?;
```

Also, if a `i64` fails to convert to a `u64`, shouldn't that be a `TypeError` (because it's a negative number), not an `Overflow` error

Reply:
IIRC we've used `ParseError` in the past, but at the end of the day, the outcome here would be that the node would panic since it's operating on corrupt chainstate.